sotodlib/site_pipeline/update_smurf_caldbs.py

```diff
@@ -109,15 +124,289 @@ def smurf_detset_info(config:str | dict,
+                pA_per_phi0 = ivas[dset]['meta']['pA_per_phi0']
+        else:
+            logger.debug("missing bias step data for %s", dset)
+
+    if rtm_bit_to_volt is None:
+        rtm_bit_to_volt = DEFAULT_RTM_BIT_TO_VOLT
+    if pA_per_phi0 is None:
+        pA_per_phi0 = DEFAULT_pA_per_phi0
+
+    # Add IV info
+    for i, cal in enumerate(cals):
+        band = am.det_info.smurf.band[i]
+        chan = am.det_info.smurf.channel[i]
+        detset = am.det_info.detset[i]
+        iva = ivas[detset]
+
+        if iva is None: # No IV analysis for this detset
+            continue
+
+        ridx = np.where(
+            (iva['bands'] == band) & (iva['channels'] == chan)
+        )[0]
+        if not ridx: # Channel doesn't exist in IV analysis
+            continue
+
+        ridx = ridx[0]
+        cal.bg = iva['bgmap'][ridx]
+        cal.polarity = iva['polarity'][ridx]
+        cal.r_n = iva['R_n'][ridx]
+        cal.p_sat = iva['p_sat'][ridx]
+    
+    obs_biases = dict(zip(am.bias_lines.vals, am.biases[:, 0] * 2*rtm_bit_to_volt))
+    bias_line_is_valid = {k: True for k in obs_biases.keys()}
+
+    # check to see if biases have changed between bias steps and obs
+    for bsa in bsas.values():
+        for bg, vb_bsa in enumerate(bsa['Vbias']):
+            bl_label = f"{bsa['meta']['stream_id']}_b{bg:0>2}"
+            if np.isnan(vb_bsa):
+                bias_line_is_valid[bl_label] = False
+                continue
+
+            if np.abs(vb_bsa - obs_biases[bl_label]) > 0.1:
+                logger.debug("bias step and obs biases don't match for %s", bl_label)
+                bias_line_is_valid[bl_label] = False
+
+    for i, cal in enumerate(cals):
+        band = am.det_info.smurf.band[i]
+        chan = am.det_info.smurf.channel[i]
+        detset = am.det_info.detset[i]
+        stream_id = am.det_info.stream_id[i]
+        bg = cal.bg
+        bsa = bsas[detset]
+
+        if bsa is None or bg == -1:
+            continue
+
+        bl_label = f'{stream_id}_b{bg:0>2}'
+        if not bias_line_is_valid[bl_label]:
+            continue
+
+        ridx = np.where(
+            (bsa['bands'] == band) & (bsa['channels'] == chan)
+        )[0]
+        if not ridx: # Channel doesn't exist in bias step analysis
+            continue
+
+        ridx = ridx[0]
+        cal.r_tes = bsa['R0'][ridx]
+        cal.r_frac = bsa['Rfrac'][ridx]
+        cal.p_bias = bsa['Pj'][ridx]
+        cal.s_i = bsa['Si'][ridx]
+        cal.phase_to_pW = pA_per_phi0 / (2*np.pi) / cal.s_i * cal.polarity
+
+    rset = core.metadata.ResultSet.from_friend(np.array(
+        [astuple(c) for c in cals], dtype=cal_dtype
+    ))
+    return rset
+
+
+def get_obs_with_detsets(ctx, detset_idx):
+    """Gets all observations with detset data"""
+    db = core.metadata.ManifestDb(detset_idx)
+    detsets = db.get_entries(['dataset'])['dataset']
+
+    obs_ids = set()
+    for dset in detsets:
+        cur = ctx.obsfiledb.conn.execute(
```

Comment:
Could we move this query into `obsfiledb`? This seems like a more general function so it'd be nice to have it available there. It's also the exact inverse of a function in there so it would match nicely.